### DISCUSSION TITLE
Is it possible to "alias" properties ?

### DISCUSSION THREAD

**Discussion by lowlighter** [colinhacks/zod#1143] [Q&A] [answered]
Hi!

Let's imagine that I have 2 APIs to interconnect, which have similar output/input but some keys are not named the same and some post-processing is done in the middle. Is it possible to "rename" on the fly the output ?

Something similar too:
```ts
const input = z.object({
  foo: z.string().alias("bar")
})

input.parse({foo: "test"}) // {bar: "test"}
```

**Comment by CanRau** [comment]
Found this looking for the same, would be great to be able rename/alias properties.
Just stumbled upon https://github.com/colinhacks/zod/issues/42 where it was mentioned ~twice but maybe was overlooked as I didn't find any real discussion about it.

I adapted a good example by @krzkaczor from the linked issue:

```ts
const envSchema = z.object({
  serverHost: z.string().alias('SERVER_HOST'),
  publicS3: z.string().alias('PUBLIC_S3_URL'),
  privateBackendURL: z.string().alias('PRIVATE_BACKEND_URL'),
})
```

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const schema = z.object( { foo: z.string() } )
type Schema = z.infer<typeof schema>
// type Schema = { foo: string }
console.log( schema.parse( { foo: 'test' } ) ) // { foo: 'test' }

const transformedSchema = schema.transform( ( { foo } ) => ( { bar: foo } ) )
type TransformedSchema = z.infer<typeof transformedSchema>
// type TransformedSchema = { bar: string }
console.log( transformedSchema.parse( { foo: 'test' } ) ) // { bar: 'test' }
```

**Reply by lowlighter** [reply]
Kind of, but I'd like to have both `foo`/`bar` supported in input schema while the output schema would only contain `bar` (which doesn't seem to be the case if I read the code you provided correctly, as `bar` wouldn't be usable in input schema)

The behaviour I'm trying to achieve is more like CLI options parsing where for example both `--file` and `-f` are the same thing, but the output would just result in a single `file` property. So basically both `{foo:"test"}`/`{bar:"test"}` would be accepted with having `{bar:"test"}` as result if `foo` is configured as an alias for `bar`

**Comment by JacobWeisenburger** [comment, marked as answer]
There are a lot of different ways to do something like this. But for brevity, here is the way I would do it.

```ts
const schema = z.union( [
    z.object( { foo: z.string() } ),
    z.object( { bar: z.string() } )
] )
type Schema = z.infer<typeof schema>
// type Schema = { foo: string } | { bar: string }
console.log( schema.parse( { foo: 'test' } ) ) // { foo: 'test' }
console.log( schema.parse( { bar: 'test' } ) ) // { bar: 'test' }

const transformedSchema = schema
    .transform( value => ( { bar: 'foo' in value ? value.foo : value.bar } ) )
type TransformedSchema = z.infer<typeof transformedSchema>
// type TransformedSchema = { bar: string }
console.log( transformedSchema.parse( { foo: 'test' } ) ) // { bar: 'test' }
console.log( transformedSchema.parse( { bar: 'test' } ) ) // { bar: 'test' }
```
